server/error_recorder.go

```diff
@@ -0,0 +1,48 @@
+package server
+
+import (
+	"net/http"
+	"runtime/debug"
+	"time"
+
+	"github.com/bytebase/bytebase/api"
+	"github.com/labstack/echo/v4"
+)
+
+func errorRecorderMiddleware(err error, s *Server, c echo.Context, e *echo.Echo) {
+	e.DefaultHTTPErrorHandler(err, c)
+
+	req := c.Request()
+	res := c.Response()
+
+	he, isHTTPError := err.(*echo.HTTPError)
+
+	if res.Status == http.StatusInternalServerError {
+		var role api.Role
+		if r, ok := c.Get(getRoleContextKey()).(api.Role); ok {
+			role = r
+		}
+
+		var stackTrace string
+		// There're basically two kinds of errors: internal HTTP errors and runtime panics.
+		// If we encounter runtime panics, we can report the stack trace for debugging.
+		if !isHTTPError || he.Internal == nil {
+			// Deal with runtime panics.
+			// Record the stack trace manually since we don't know where it'll happen.
```

Comment:
This comment is a bit confusing since the panic has already happened.

Reply:
How about

```suggestion
			// Deal with runtime panics.
			// Since we don't know where future panics will occur, we will manually record the stack trace here
			// so that we can handle all possible panics.
```

Reply:
Still confused, I thought the order is that we first encountered panic and then we got here with 500 internal. What does future panics refer to?